File: algorithm/leetcode1218.longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
    	unordered_map<int,int> m;//等差数列结尾的数，最大长度
    	int maxlen = 1;
    	for(int i = 0, prev; i < arr.size(); ++i)
    	{
    		prev = arr[i]-difference;
    		if(m.count(prev))//前一个数
    			m[arr[i]] = m[prev] + 1;
    		else
    			m[arr[i]] = 1;
    		maxlen = max(maxlen, m[arr[i]]);
    	}
    	return maxlen;
    }
};

File: algorithm/Python3_version/leetcode1218.longest-arithmetic-subsequence-of-given-difference.py
class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        d = {}
        maxlen = 1
        for i in range(len(arr)):
            if arr[i]-difference in d:
                d[arr[i]] = d[arr[i]-difference] + 1
            else:
                d[arr[i]] = 1
            maxlen = max(maxlen, d[arr[i]])
        return maxlen